### PR TITLE
Added section for contributing to Magics

### PR DESCRIPTION
**Pull request by Manvi-Agrawal**
This is as per the following comments : https://github.com/microsoft/QuantumKatas/issues/166#issuecomment-827118778, https://github.com/microsoft/QuantumKatas/issues/166#issuecomment-783879320 and https://github.com/microsoft/QuantumKatas/issues/166#issuecomment-802951251

### CODE REVIEW

**Comment by Manvi-Agrawal** [comment]
Hi @tcNickolas. Changed the instructions in [Contributing Guide](https://github.com/microsoft/QuantumKatas/blob/main/.github/CONTRIBUTING.md) a bit to avoid skipping the auto-loading of projects by IQ# kernel. I think this is a slightly better technique to test the changes locally since the end goal is to ensure IQ# is successfully auto-loading the projects. Also, this technique allows to manipulate cache assembly in a bit more convenient way from the `KataMagic` and `CheckKataMagic`, should we decide to do in future.

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -159,6 +160,47 @@ This can happen for several reasons:
 To exclude a task from validation, open the corresponding Jupyter notebook and choose ```View -> Cell Toolbar -> Tags``` to see and edit the tags for each cell. Add the tag that is the most fitting description of the failure cause to the cell. 
 After you are done with editing the notebook, choose ```View -> Cell Toolbar -> None``` to turn off tags editing view for the subsequent users of this notebook. Finally, save the notebook.
 
+#### Validate changes to `%kata` magic and `%check-kata` magic on local machine
+1. Add logging statements to the [`KataMagic`](./utilities/Microsoft.Quantum.Katas/KataMagic.cs) and [`CheckKataMagic`](./utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs).
```
Comment (tcNickolas) [inline .github/CONTRIBUTING.md:164]: Since this file is not in the repo root but in `.github` folder, the relative links need to go one level up, starting with `./../` rather than just `./` (later in the change as well)

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -159,6 +160,47 @@ This can happen for several reasons:
...
+   ```
+   Logger.LogDebug($"Value : {val}");
+   ```
+2. Generate a custom nuget package
+   1. Build the Microsoft.Quantum.Katas project to produce a NuGet package to get a version other than 1.0.0 by adding the following code in [this file](./utilities/Microsoft.Quantum.Katas/Microsoft.Quantum.Katas.csproj) by adding the following code
+   ```
+   <RootNameSpace>Microsoft.Quantum.Katas</RootNameSpace>
+   <Version>version-number</Version> ).
```
Comment (tcNickolas) [inline .github/CONTRIBUTING.md:172]:
```suggestion
   <Version>version-number</Version>
```

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -159,6 +160,47 @@ This can happen for several reasons:
...
+   <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+   ```
+   2. Copy the generated .nupkg file to the folder in which your project resides (like PhaseEstimation kata) and remove the following line from your project's project file.
+   ```
+   <PackageReference Include="Microsoft.Quantum.Katas" Version="0.16.2104138035" />
```
Comment (tcNickolas) [inline .github/CONTRIBUTING.md:177]: If we mention the package version here, we'll need to update version update script to cover this line. (You'll see that [right now](https://github.com/microsoft/QuantumKatas/blob/35c153fe5b4901b0a822813ddaa77d4aeedcf165/scripts/Update-QDKVersion.ps1#L94) we update another line in this script, but not this one). I'd suggest to avoid this (especially since here the version number can be pretty much anything) and rephrase this as "Remove the package reference to Microsoft.Quantum.Katas from your .csproj file"

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -159,6 +160,47 @@ This can happen for several reasons:
...
+   <?xml version="1.0" encoding="utf-8"?>
+   <configuration>
+     <packageSources>
+       <add key="Local Folder" value="." />
+     </packageSources>
+   </configuration>
+   ```
+3. To observe your changes in console window where Jupyter Notebook was launched.
+   1. Set the environment variable `IQSHARP_LOG_LEVEL='Debug'`.
+   > Note : The environment variable is case-sensitive.
```
Comment (tcNickolas) [inline .github/CONTRIBUTING.md:190]: I think the notes and the samples here (and above as well) should be indented more to render as part of second-level list item rather than of first-level. This should make it more readable

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -159,6 +160,47 @@ This can happen for several reasons:
...
+   ```
+   $ start jupyter notebook
+   ```
+   > This would launch Jupyter Notebooks server in a new window which would be
+   > helpful in debugging the magics since logging is a bit noisy.
+
+4. Changes to be done in jupyter notebook 
+   1. Use `%package` magic to load the newly built nuget package `Microsoft.Quantum.Katas` by adding the following code
+   ```
+   %package Microsoft.Quantum.Katas::<package-version>
```
Comment (tcNickolas) [inline .github/CONTRIBUTING.md:201]: Do we need this line, or would it work if we just change the version in the `<PackageReference Include="Microsoft.Quantum.Katas" Version="0.16.2104138035" />` dependency in the .csproj file? Trying to figure out how to minimize the number of steps

**Review by tcNickolas — CHANGES_REQUESTED** [review]
I did a quick pass, I still have to try these steps myself to make sure they work, but I left some comments independently of that.

Thank you!

File: .github/CONTRIBUTING.md (hunk 5 above)
Comment (Manvi-Agrawal) [reply .github/CONTRIBUTING.md:201]: I initially tried that approach only but the IQ# kernel for no particular reason loaded the katas package of the same version as the IQ# kernel. I think this might be due to the fact that IQ# kernel would attempt to load dependencies of the same version as the kernel if <IQSharpLoadAutomatically> property is set to true.

File: .github/CONTRIBUTING.md (hunk 4 above)
Comment (Manvi-Agrawal) [reply .github/CONTRIBUTING.md:190]: Done

File: .github/CONTRIBUTING.md (hunk 1 above)
Comment (Manvi-Agrawal) [reply .github/CONTRIBUTING.md:164]: Done

File: .github/CONTRIBUTING.md (hunk 3 above)
Comment (Manvi-Agrawal) [reply .github/CONTRIBUTING.md:177]: Done

**Comment by Manvi-Agrawal** [comment]
Hi @tcNickolas. I have made the desired changes.

And as for this one :
> Do we need this line, or would it work if we just change the version in the `<PackageReference Include="Microsoft.Quantum.Katas" Version="0.16.2104138035" />` dependency in the .csproj file? Trying to figure out how to minimize the number of steps.

I initially tried that approach only but the IQ# kernel for no particular reason loaded the katas package of the same version as the IQ# kernel. I think this might be due to the fact that IQ# kernel would attempt to load dependencies of the same version as the kernel if `IQSharpLoadAutomatically` is set to true.

**Comment by Manvi-Agrawal** [comment]
Hi @tcNickolas . I noticed a very strange phenomenon after qdk-update to the the version `0.16.2104138035`

if we do the following in NuGet.config as per step 2.3
```
<add key="Local Folder" value="." />
```
- In the version `0.16.2104138035`
Custom nuget now needs to be placed in the `QuantumKatas` folder(ie the same folder as that of `NuGet.config` file) 

- With early versions : `0.15.2101125897` and  `0.15.2103133969`
 We needed to move the generated nuget to the project in which the project resided.

Maybe @rmshaffer could help in confirming if changes in IQ# kernel are done or there is some problem with my local :-(

**Comment by tcNickolas** [comment]
Sadly, Ryan is no longer with our team, so @anjbur is picking up IQ# ownership.
@anjbur, were there any changes that could cause this behavior change in the past month?

Generally I would suggest recommending the person working with the custom Microsoft.Quantum.Katas package to put both NuGet.config and the custom package to the folder of the project they're using as a test, rather than in folders closer to root. That's what I do myself, to limit the scope of the custom settings and to make it harder to get unintended consequences when working with other katas.

**Comment by Manvi-Agrawal** [comment]
Apologies for my previous comment. Initially when I was following step 3 as mentioned [here](https://github.com/microsoft/QuantumKatas/issues/166#issuecomment-783879320), I changed the `NuGet.config` file present in the root folder of the `QuantumKatas` and added `<add key="Local Folder value="." />` and things seemed to go pretty well in `0.15.2101125897` and `0.15.2103133969` until the next Qdk-update to version `0.16.2104138035`.

As of now, these instructions work perfectly on my local machine. My only concern is that if someone would attempt to run the run the validate notebook script for the first time after creating the custom package, then the check would fail at `%package Microsoft.Quantum.Katas::<version-number>` since we explicitly clear the NuGet cache before validating the change. So, I could possibly update the instructions to change [the line here](https://github.com/microsoft/QuantumKatas/blob/main/scripts/validate-notebooks.ps1#L88) to `<add key=""Local Folder"" value=""."" />` . However if someone is likely to introduce a major change as in [Issue#580](https://github.com/microsoft/QuantumKatas/issues/580), I think they would be doing changes to the `KataMagic` first and then doing changes to the `CheckKataMagic`, so this situation won't arise.

**Review by tcNickolas — APPROVED** [review]
I finally got the time to sit down and go through all the debug steps - they look great, I think the only thing I didn't need was the root namespace. I did a bit of editing to call out some things that got me confused (for example, that removing package reference to M.Q.Katas will break the build but that's ok).

Thank you!
